toplevel/toploop.ml

```diff
@@ -206,21 +218,42 @@ let loop ppf =
   run_hooks After_setup;
   load_ocamlinit ppf;
   while true do
-    let snap = Btype.snapshot () in
+    let snap = ref (Btype.snapshot ()) in
     try
       Lexing.flush_input lb;
       (* Reset the phrase buffer when we flush the lexing buffer. *)
       Buffer.reset phrase_buffer;
       Location.reset();
-      Warnings.reset_fatal ();
       first_line := true;
-      let phr = try !parse_toplevel_phrase lb with Exit -> raise PPerror in
-      let phr = preprocess_phrase ppf phr  in
-      Env.reset_cache_toplevel ();
-      ignore(execute_phrase true ppf phr)
+      let rec get_phrases phrs =
+        match !parse_toplevel_phrase lb with
+        | phr ->
+          if is_blank lb then
+            List.rev (phr :: phrs)
+          else
+            get_phrases (phr :: phrs)
+        | exception Exit -> raise PPerror
+        | exception e -> Location.report_exception ppf e; []
+      in
+      let phrs = get_phrases [] in
+      let saved_report_printer = !Location.report_printer in
+      let nonfirst_report_printer = fun () -> Location.batch_mode_printer in
+      let snap = ref (Btype.snapshot ()) in
+      let finally () = Location.report_printer := saved_report_printer in
+      Fun.protect ~finally begin fun () ->
+        let execute i phr =
+          if i > 0 then Location.report_printer := nonfirst_report_printer;
+          snap := Btype.snapshot ();
+          Warnings.reset_fatal ();
+          let phr = preprocess_phrase ppf phr in
+          Env.reset_cache_toplevel ();
+          ignore(execute_phrase true ppf phr)
+        in
+        List.iteri execute phrs
+      end
```

Comment:
I read the new logic in this function and I believe that it is correct (modulo comment about shadowing `snap`), but it is a beast and not so easy to understand -- I think I would be lost as to what we are doing and why if I wasn't reading the code in the context of the PR itself.

Could you:

1. Move the logic after `get_phrases` to a helper function `execute_phrases`, to make the `loop` function itself shorter and spread the complexity around?
2. add a comment that explains why we change the location printer temporarily?

Reply:
I've moved the parsing and the processing to their own auxiliary functions and fixed a bug with very long lines that I noticed while doing that.

Reply:
I've also added a comment to explain the prompt business.